troubleshooting/tools/log-loss-test-framework/validator/validate.go

```diff
@@ -0,0 +1,272 @@
+	var forwardToken *string
+	var input *cloudwatchlogs.GetLogEventsInput
+	cwRecoredCounter := 0
+
+	// Returns all log events from a CloudWatch log group with the given log stream.
+	// This approach utilizes NextForwardToken to pull all log events from the CloudWatch log group.
+	for {
+		if forwardToken == nil {
+			input = &cloudwatchlogs.GetLogEventsInput{
+				LogGroupName:  aws.String(logGroup),
+				LogStreamName: aws.String(logStream),
+				StartFromHead: aws.Bool(true),
+			}
+		} else {
+			input = &cloudwatchlogs.GetLogEventsInput{
+				LogGroupName:  aws.String(logGroup),
+				LogStreamName: aws.String(logStream),
+				NextToken:     forwardToken,
+				StartFromHead: aws.Bool(true),
+			}
+		}
+
+		response, err := cwClient.GetLogEvents(input)
+		for err != nil {
+			// retry for throttling exception
+			if strings.Contains(err.Error(), "ThrottlingException: Rate exceeded") {
+				time.Sleep(1 * time.Second)
+				response, err = cwClient.GetLogEvents(input)
+			} else {
+				exitErrorf("[TEST FAILURE] Error occured to get the log events from log group: %q., %v", logGroup, err)
+			}
+		}
+
+		for _, event := range response.Events {
+			log := aws.StringValue(event.Message)
+
+			// First 8 char is the unique record ID
+			recordId := log[:8]
+			cwRecoredCounter += 1
+			if _, ok := inputMap[recordId]; ok {
+				// Setting true to indicate that this record was found in the destination
+				inputMap[recordId] = true
+			}
+		}
+
+		// Same NextForwardToken will be returned if we reach the end of the log stream
+		if aws.StringValue(response.NextForwardToken) == aws.StringValue(forwardToken) {
+			break
+		}
+
+		forwardToken = response.NextForwardToken
+	}
+
+	return cwRecoredCounter, inputMap
+}
+
+func printResults(totalInputRecord int, totalRecordFound int, recordMap map[string]bool, logGroupName string, logStreamName string, metadata ECSTaskMetadata) {
+	uniqueRecordFound := 0
+	// Count how many unique records were found in the destination
+	for _, v := range recordMap {
+		if v {
+			uniqueRecordFound++
+		}
+	}
+
+	totalSizeInMB := os.Getenv(envTotalSizeInMB)
+
+	logSizeInKB:= os.Getenv(envLogSizeInKB)
+
+	throughputInKB := os.Getenv(envThroughputInKB)
+
+	testName := os.Getenv(envTestName)
+
+	var hasLogLoss string
+
+	if totalInputRecord != uniqueRecordFound {
+		hasLogLoss = "🚨"
+	} else {
+		hasLogLoss = "✅"
+	}
+
+	// if there are many test cases to run
+	// perform CW insights query for test name
+	// output is comma delimited, output insights query result as CSV
+	// simple python code can parse the CSV
+	fmt.Printf("%s %s - %s, percent lost, %d, number_lost, %d, total_input_record, %d, duplicates, %d, group=%s stream=%s TOTAL_SIZE_IN_MB=%s, SIZE_IN_MB=%s, THROUGHPUT_IN_KB=%s, %s, %s, %s, %s, %s",
+			   testName, hasLogLoss, throughputInKB, (totalInputRecord-uniqueRecordFound)*100/totalInputRecord, totalInputRecord-uniqueRecordFound, totalInputRecord, totalRecordFound - uniqueRecordFound,
```

Comment:
This looks good.

Reply:
I should've put the buffer size in it, but oh well, thats in the test name right now